---
title: "NINA Analysis: Pair-wise Comaprisons Against the Control"
author: "Javier"
date: "2025-04-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Ease of life

include = FALSE prevents code and results from appearing in the finished file echo = FALSE prevents code, but not the results from appearing in the finished file message = FALSE prevents messages that are generated by code from appearing in the finished file warning = FALSE prevents warnings that are generated by code from appearing in the finished. fig.cap = "..." adds a caption to graphical results.

# Package Installation

The following code is used to download all libraries necessary for this analysis. Differential gene expression is performed using DEseq2 with plotting done with ggplot2 and ggVennDiagram.

```{r, echo = FALSE, message = FALSE}
### REQUIRED PACKAGES ###

# install bioconductor packages
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
if (!requireNamespace("DESeq2", quietly = TRUE))
  BiocManager::install("DESeq2")
if (!requireNamespace("topGO", quietly = TRUE))
  BiocManager::install("topGO")
if (!requireNamespace("clusterProfiler", quietly = TRUE))
  BiocManager::install("clusterProfiler")
if (!requireNamespace("KEGGREST", quietly = TRUE))
  BiocManager::install("KEGGREST")
# install CRAN packages
if (!requireNamespace("ggVennDiagram", quietly = TRUE))
  install.packages("ggVennDiagram")
if (!requireNamespace("ggplot2", quietly = TRUE))
  install.packages("ggplot2")
if (!requireNamespace("dplyr", quietly = TRUE))
  install.packages("dplyr")
if (!requireNamespace("pheatmap", quietly = TRUE))
  install.packages("pheatmap")
if (!requireNamespace("tidyr", quietly = TRUE))
  install.packages("tidyr")
if (!requireNamespace("enrichplot", quietly = TRUE))
  install.packages("enrichplot")
if (!requireNamespace("ggrepel", quietly = TRUE))
  install.packages("ggrepel")
if (requireNamespace("progress", quietly = TRUE))
  install.packages("progress")
if (!requireNamespace("cowplot", quietly = TRUE))
  install.packages("cowplot")
if (!requireNamespace("gt", quietly = TRUE))
  install.packages("gt")
if (!requireNamespace("webshot2", quietly = TRUE))
  install.packages("webshot2")
```

# Loading Libraries

This section of code loads the necessary libraries.

```{r, echo = FALSE, message = FALSE}
# load libraries
library("DESeq2")
library("ggplot2")
library("ggVennDiagram")
library("dplyr")
library("pheatmap")
library("topGO")
library("clusterProfiler")
library("tidyr")
library("enrichplot")
library("topGO")
library("ggrepel")
library("KEGGREST")
library("progress")
library("cowplot")
library("gt")
library("webshot2")
```

# Data Import

The following code is used to import the data for DEG analysis. The text file produced from feature counts is imported as the raw table. The table is cleaned of unnecessary rows, ie meta-data.

The row names then need to be set for the genes list or else DESeq2 will not work. It only takes numbers as input and extra information not in the form of a column or row name will result in an error message.

The first column contains a set of numbers that corresponds to the row number and needs to be removed before the gene list can be assigned as the row name.

Coldata refers to column data, ie. the meta data needed to perform analysis. This table has information about each analyzed fastq file and the treatment it received. In my case this was control, nina treated at 24 and 48 hours, and treatment with bacterial cells for 24 and 48 hours.

Similarly to the featurecounts data, the column-data needed reformatting with row names corresponding to each fastq file in order.

The final line is a way to confirm that the columns in the feature count table correspond with the rows of the coldata table.

```{r, echo = FALSE}
### IMPORTING DATA ###

# analyzing count matrix
featurecounts_raw <- read.delim("feature_counts.txt",
                                sep = "\t",
                                header = TRUE) # import feature_counts
cleanfeatures <- featurecounts_raw[-c(2, 3, 4, 5, 6)] # cleans up columns

# set row names
counts <- cleanfeatures[, -1] # removes the first column
rownames(counts) <- cleanfeatures[, 1] # sets row names for counts

# import column data
coldata <- read.csv("coldata.csv", header = TRUE)

# set row names for coldata
rownames(coldata) <- coldata$X # the name of the column is X
```

# Column and Row Sanity Check

The following snippet of codes checks to see if columns and rows match for the coldata and cleaned and formatted counts matrix. If the following code prints FALSE, then go back and trouble shoot.

```{r}
# test to see if columns and rows match
all(rownames(coldata) == colnames(counts)) # should be true
```

# DEG Analysis

The following snippet of code performs the DEG analysis.

The first snippet creates an object called dds, ie. DESeqDataSet. The parameters include the data, in this case it was the counts object we created in the last code block with the gene list as the row names. The column data is the coldata object we created and reformatted. Finally, the design includes the column name for the coldata object holding our experimental design, ie. treatment type.

The second line of code sets the control group as the reference for comparison.

```{r, echo = FALSE}
### DEG ANALYSIS ###

# creates a DeseqDataSet or DDS
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = coldata,
                              design = ~ Treatment)

# sets the comparisons, in this case the coldata column name is Treatment,
# the reference is named control
dds$Treatment <- relevel(dds$Treatment, ref = "control")

# no clue what this does, manual said to tho
# prints the dataframe i suppose?
dds

# performs DEG
dds <- DESeq(dds)

# creates a results dataframe
results <- results(dds)
```

# Sanity Check for Sucessful DEG Analysis

The following code prints all results, double check and make sure the proper comparisons were made and trouble shoot if need be.

```{r}
# prints results
results

# prints summary of results
summary(results)

# prints pairwise comparisons,useful double check
resultsNames(dds)
```

# PCA Plot

The following snippet of code is crucial for testing the quality of our data.

The first line of code creates an object called vsd. I do not remember why I chose vsd, I think it might have been variance stablizied data but I am not sure. In any case, the line of code is taking the dds object with our differential expression analysis data and applying the method vst to it. vst is one of the variance stablizing functions provided by DESeq2 that allows for the creation of PCA plots. By default, the vst method used the top 500 genes in terms of variance for the analysis. You can modify this using the ntop = foo option, The blind = TRUE option is important, as it allows you to see the variance across groups.

The next line of code creates an object called pca_data that stored the output from the plotPCA method provided by DESEq2. It uses our vsd object, intgroup refers to experimental design, and returnData = TRUE does something important I presume.

Finally, the next line of code creates an object percent_var that stores the percent variance extracted from the pcr_data.

```{r}
### RUN THIS FIRST AFTER DEG TO ENSURE QUALITY ###

# creates an output directory
dir.create("DEG_results", showWarnings = FALSE)

# VST transformation
vsd <- vst(dds, blind = TRUE)
pca_data <- plotPCA(vsd, intgroup = c("Treatment"), returnData = TRUE)
percent_var <- round(100 * attr(pca_data, "percentVar"))

# PCA plotting using ggplot2
pca <- ggplot(pca_data, aes(PC1, PC2, color = Treatment)) +
  geom_point(size = 3, alpha = 0.8) +
  xlab(paste0("PC1: ", percent_var[1], "% variance")) +
  ylab(paste0("PC2: ", percent_var[2], "% variance")) +
  coord_fixed() +
  scale_color_brewer(palette = "Set1") +
  theme_minimal(base_size = 14) +
  theme(
    legend.position = "right",
    legend.title = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    axis.text = element_text(color = "black"),
    panel.grid.major = element_line(color = "gray90"),
    panel.grid.minor = element_blank()
  )

# save plot as a png
ggsave("DEG_results/pca_plot.png", plot = pca, width = 6, height = 6, dpi = 300)

# print to r markdown
pca
```

# Venn Diagrams

Since our quality was all good, we can start creating some plots to represent our data.

```{r}
# creates an output directory
dir.create("DEG_results", showWarnings = FALSE)

# list to store genes
comparisons <- resultsNames(dds)[-1] # removes the first weird column
all_degs <- list()

# loop through pair-wise comparison
for (comp in comparisons) {
  res <- results(dds, name = comp) #saves results for each comparison
  sig <- res[which(res$padj < 0.05 & !is.na(res$padj)), ] # filter for p and NA

  up <- rownames(sig[sig$log2FoldChange > 1.5, ]) # filter for up-reg 1.5
  down <- rownames(sig[sig$log2FoldChange < -1.5, ]) # filter for down-red -1.5

  all_degs[[comp]] <- up # compile up-reg genes
  all_degs[[comp]] <- down # compile down-reg genes
}

# todos los genes arriba y abajo
all_venn <- ggVennDiagram(all_degs, label_alpha = 0, edge_size = 0.5,
                          label = "count",
                          label_size = 5,
                          category.names = c("A", # cells vs control 24H
                                             "B", # cells vs control 48h
                                             "C", # supernatant vs c 24h
                                             "D")) + # super vs c 48
  scale_fill_distiller(palette = "Spectral") +
  labs(title = "") +
  theme(text = element_text(size = 10),
        plot.title = element_text(size = 14, hjust = 0.5, face = "bold"))

# fixes the long labels
fixed_venn <- all_venn + scale_x_continuous(expand = expansion(mult = 0.2))

ggsave("DEG_results/all_DEGs_venn.png",
       plot = fixed_venn, dpi = 600, width = 10, height = 8)
# plot to markdown
fixed_venn
```

# Extraction of Indiviudal Comparisons

Some of the following analyses involve the inividual pair-wise comparisons, so in this section I am extracting them and stored them as dataframes.

```{r, echo = FALSE}
# store comparison names
cell_24_control <- resultsNames(dds)[-c(1,3,4,5)] # this part is pretty confusing
cell_48_control <- resultsNames(dds)[-c(1,2,4,5)] # what i am doing is removing..
nina_24_control <- resultsNames(dds)[-c(1,2,3,5)] # the different indices to leave..
nina_48_control <- resultsNames(dds)[-c(1,2,3,4)] # the one I am interested in

#saves results for each comparison
resC24 <- results(dds, name = cell_24_control)
resC48 <- results(dds, name = cell_48_control) 
resN24 <- results(dds, name = nina_24_control)
resN48 <- results(dds, name = nina_48_control)

#filter the data for NA values
resC24_df <- as.data.frame(resC24) %>% # pipe operator, the same as | foo | in bash
  filter(!is.na(log2FoldChange), !is.na(padj))
resC48_df <- as.data.frame(resC48) %>%
  filter(!is.na(log2FoldChange), !is.na(padj))
resN24_df <- as.data.frame(resN24) %>%
  filter(!is.na(log2FoldChange), !is.na(padj))
resN48_df <- as.data.frame(resN48) %>%
  filter(!is.na(log2FoldChange), !is.na(padj))

# add columns for up, down, and no-sig
resC24_df$diffexpress <- "NO" # initializes column with no as default
resC24_df$diffexpress[resC24_df$padj < 0.05 & resC24_df$log2FoldChange > 1.5] <- "UP"
resC24_df$diffexpress[resC24_df$padj < 0.05 & resC24_df$log2FoldChange < -1.5] <- "DOWN"

resC48_df$diffexpress <- "NO" # initializes column with no as default
resC48_df$diffexpress[resC48_df$padj < 0.05 & resC48_df$log2FoldChange > 1.5] <- "UP"
resC48_df$diffexpress[resC48_df$padj < 0.05 & resC48_df$log2FoldChange < -1.5] <- "DOWN"

resN24_df$diffexpress <- "NO" # initializes column with no as default
resN24_df$diffexpress[resN24_df$padj < 0.05 & resN24_df$log2FoldChange > 1.5] <- "UP"
resN24_df$diffexpress[resN24_df$padj < 0.05 & resN24_df$log2FoldChange < -1.5] <- "DOWN"

resN48_df$diffexpress <- "NO" # initializes column with no as default
resN48_df$diffexpress[resN48_df$padj < 0.05 & resN48_df$log2FoldChange > 1.5] <- "UP"
resN48_df$diffexpress[resN48_df$padj < 0.05 & resN48_df$log2FoldChange < -1.5] <- "DOWN"

# convert row names to columns
resC24_df$gene_id <- rownames(resC24_df)
resC48_df$gene_id <- rownames(resC48_df)
resN24_df$gene_id <- rownames(resN24_df)
resN48_df$gene_id <- rownames(resN48_df)
```

# Adding DIAMOND hit descriptions to label genes

A nice trouble shooting snippet setdiff(resC24_df$gene_id, diamond_hits$qseqid)
Prints out any non-matching IDs. For some reason none of mine matched at first, which was caused by a white space issue.

```{r}
# reading in our annotation file
#diamond_hits <- read.delim("filtered.tsv", sep = "\t", header = TRUE)

# trimming white space off the columns
diamond_hits$qseqid <- trimws(diamond_hits$qseqid)
diamond_hits$sseqid <- trimws(diamond_hits$sseqid)
diamond_hits$stitle <- trimws(diamond_hits$stitle)

# function to add annotations to results dataframe
add_diamond_annotations <- function(res_df, diamond_df) {
  merge(res_df, diamond_df[, c("qseqid", "sseqid", "stitle")],
        by.x = "gene_id", by.y = "qseqid", all.x = TRUE) # uses a left join (all.x)
}

# running our function for each data frame
DIAMOND_resC24_df <- add_diamond_annotations(resC24_df, diamond_hits)
DIAMOND_resC48_df <- add_diamond_annotations(resC48_df, diamond_hits)
DIAMOND_resN24_df <- add_diamond_annotations(resN24_df, diamond_hits)
DIAMOND_resN48_df <- add_diamond_annotations(resN48_df, diamond_hits)

# adding a new column with only upregulated genes for labeling
add_delabel_top50 <- function(df) {
  
  # selects our significant genes using both values
  sig_genes <- df %>%
    filter(abs(log2FoldChange) > 1.5, padj < 0.05)
  
  # selects the top 25 by p-value
  top50 <- sig_genes %>%
    arrange(padj) %>%
    slice_head(n = 10)
  
  # add delabel column, label only top 50 genes
  df$delabel <- ifelse(df$sseqid %in% top50$sseqid, df$sseqid, NA)
  
  return(df)
}

# Apply to your dataframes
DIAMOND_resC24_df <- add_delabel_top50(DIAMOND_resC24_df)
DIAMOND_resC48_df <- add_delabel_top50(DIAMOND_resC48_df)
DIAMOND_resN24_df <- add_delabel_top50(DIAMOND_resN24_df)
DIAMOND_resN48_df <- add_delabel_top50(DIAMOND_resN48_df)
```


# Volcano Plots

These are classic for DEG analysis, Kaleb said to make one too so here goes nothing. I followed a tutorial from biostatsquid and the plots are made with ggplot2 and dply for data wrangling. Very useful tutorial btw, I learned that ggplot2 adds elements as you place them in the script. Like the layers of an onion.

Also, if you do ggplot2(foo) it works differently than ggplot2(data = foo).

```{r}
# creates a basic volcano plot as a starting point
volc_cells_24 <- ggplot(data = DIAMOND_resC24_df, 
                        aes(x = log2FoldChange, 
                            y = -log10(padj), 
                            col = diffexpress)) +
  geom_vline(xintercept = c(-1.5, 1.5), col = "gray", linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = "dashed") +
  geom_point(size = 2) +
  scale_color_manual(values = c("blue","grey","red"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, 40), xlim = c(-10, 10)) +
  labs(color = "Legend", 
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"p-value")) +
  ggtitle("Cells vs. Control 24H") +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

volc_cells_48 <- ggplot(data = resC48_df, 
                        aes(x = log2FoldChange, 
                            y = -log10(padj), 
                            col = diffexpress)) +
  geom_vline(xintercept = c(-1.5, 1.5), col = "gray", linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = "dashed") +
  geom_point(size = 2) +
  scale_color_manual(values = c("blue","grey","red"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, 40), xlim = c(-10, 10)) +
  labs(color = "Legend", 
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"p-value")) +
  ggtitle("Cells vs. Control 48H") +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

volc_NINA_24 <- ggplot(data = resN24_df, 
                       aes(x = log2FoldChange, 
                           y = -log10(padj), 
                           col = diffexpress)) +
  geom_vline(xintercept = c(-1.5, 1.5), col = "gray", linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = "dashed") +
  geom_point(size = 2) +
  scale_color_manual(values = c("blue","grey","red"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, 40), xlim = c(-10, 10)) +
  labs(color = "Legend", 
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"p-value")) +
  ggtitle("Supernatant vs. Control 24H") +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

volc_NINA_48 <- ggplot(data = resN48_df, 
                       aes(x = log2FoldChange, y = -log10(padj), col = diffexpress)) +
  geom_vline(xintercept = c(-1.5, 1.5), col = "gray", linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = "dashed") +
  geom_point(size = 2) +
  scale_color_manual(values = c("blue","grey","red"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, 40), xlim = c(-10, 10)) +
  labs(color = "Legend", 
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"p-value")) +
  ggtitle("Supernatant vs. Control 48H") +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

# save
#ggsave("thesis/cellsv24.png", plot = volc_cells_24, dpi = 600, width = 10, height = 8)
#ggsave("thesis/cellsv48.png", plot = volc_cells_48, dpi = 600, width = 10, height = 8)
#ggsave("thesis/superv24.png", plot = volc_NINA_24, dpi = 600, width = 10, height = 8)
#ggsave("thesis/superv48.png", plot = volc_NINA_48, dpi = 600, width = 10, height = 8)

# plot to markdown
volc_cells_24
volc_cells_48
volc_NINA_24
volc_NINA_48
```
# Merging plots into one multipanel figure

This is done using cowplot

```{r}
# cleanup legends
volc_cells_24 <- volc_cells_24 + theme(legend.position = "none")
volc_cells_48 <- volc_cells_48 + theme(legend.position = "none")
volc_NINA_24 <- volc_NINA_24  + theme(legend.position = "none")
volc_NINA_48 <- volc_NINA_48 + theme(legend.position = "none")

comb_volc_first <- plot_grid(volc_cells_24, volc_cells_48, volc_NINA_24, volc_NINA_48, 
                             labels = c("A","B","C","D"),
                             ncol = 2,
                             align = "hv")
# save plot
ggsave("DEG_results/final_volcano_clean.png", 
       plot = comb_volc_first, dpi = 600, width = 10, height = 8)

# plot to markdown
comb_volc_first
```


# Loading in eggnog mapper annotations

```{r}
# Load in our data
annotation_data <- read.delim("dups_removed_viridae.tsv", sep = "\t", header = TRUE, check.names = FALSE)

# check.names = FALSE helps with R putting X in front of columns that start with a number

# colnames(annotation_data)[1] <- gsub("^X\\.", "", colnames(annotation_data)[1])
# run this if the column names import all funny
```

# Gonna give GO analysis another GO badum tss, prep work part 1

Making our gene universe. Gotta filter out any genes that do not have any go terms or it'll wash out our analysis.

```{r}
# Going to create our gene universe from the eggnog mapper output
go_terms <- annotation_data[c(1,10)]
go_terms <- go_terms[go_terms$GOs != "-",]

write.table(go_terms, file = "go_terms.tsv", sep = "\t", quote = FALSE, row.names = FALSE)

gene2goMappings <- readMappings(file = "go_terms.tsv", sep = "\t", IDsep = ",")
```

# I think we ready

I am merging the go term data frame with the results frame from the DESeq analysis. Then I am removing any genes that do not have a go term to not wash out the analysis.

```{r}
# making a frame with only the genes with GO terms
go_C24 <- merge(resC24_df, go_terms[, c("query", "GOs")], by.x = "gene_id", by.y = "query", all.x = TRUE)
go_C48 <-merge(resC48_df, go_terms[, c("query", "GOs")], by.x = "gene_id", by.y = "query", all.x = TRUE)
go_N24 <- merge(resN24_df, go_terms[, c("query", "GOs")], by.x = "gene_id", by.y = "query", all.x = TRUE)
go_N48 <- merge(resN48_df, go_terms[, c("query", "GOs")], by.x = "gene_id", by.y = "query", all.x = TRUE)

# filter out any genes with no GO term
go_C24 <- go_C24[!is.na(go_C24$GOs),]
go_C48 <- go_C48[!is.na(go_C48$GOs),]
go_N24 <- go_N24[!is.na(go_N24$GOs),]
go_N48 <- go_N48[!is.na(go_N48$GOs),]

# extract genes with logfold of abs 2
go_C24 <- go_C24 %>%
  filter(abs(log2FoldChange) > 1.5)

go_C48 <- go_C48 %>%
  filter(abs(log2FoldChange) > 1.5)

go_N24 <- go_N24 %>%
  filter(abs(log2FoldChange) > 1.5)

go_N48 <- go_N48 %>%
  filter(abs(log2FoldChange) > 1.5)

# Extract gene IDs and p-values
Cp24 <- dplyr::select(go_C24, gene_id,padj)
Cp48 <- dplyr::select(go_C48, gene_id,padj)
Np24 <- dplyr::select(go_N24, gene_id,padj)
Np48 <- dplyr::select(go_N48, gene_id,padj)

# sort by p-value
sorted_Cp24 <- Cp24[order(Cp24$padj),]
sorted_Cp48 <- Cp48[order(Cp48$padj),]
sorted_Np24 <- Np24[order(Np24$padj),]
sorted_Np48 <- Np48[order(Np48$padj),]

# convert to topGOs genelist format
topgo_Cp24 <- sorted_Cp24$padj
topgo_Cp48 <- sorted_Cp48$padj
topgo_Np24 <- sorted_Np24$padj
topgo_Np48 <- sorted_Np48$padj

# name the topgo vector
names(topgo_Cp24) <- sorted_Cp24$gene_id
names(topgo_Cp48) <- sorted_Cp48$gene_id
names(topgo_Np24) <- sorted_Np24$gene_id
names(topgo_Np48) <- sorted_Np48$gene_id

# filtering value
padj_cut <- 0.05

# functions for filtering significant genes
top5C24_func <- function(x) {
  return(x <= padj_cut)
}

top5C48_func <- function(x) {
  return(x <= padj_cut)
}

top5N24_func <- function(x) {
  return(x <= padj_cut)
}

top5N48_func <- function(x) {
  return(x <= padj_cut)
}
```

# Creating GO data objects for enrichment tests

```{r, echo=FALSE,message=FALSE}
# creating GO data objects for molecular function
Cp24_go_mf <- new("topGOdata", 
               ontology = "MF", 
               allGenes = topgo_Cp24, 
               annot = annFUN.gene2GO, 
               geneSel = top5C24_func,
               gene2GO = gene2goMappings)

Cp48_go_mf <- new("topGOdata",
               ontology = "MF",
               allGenes = topgo_Cp48,
               annot = annFUN.gene2GO,
               geneSel = top5C48_func,
               gene2GO = gene2goMappings)

Np24_go_mf <- new("topGOdata",
               ontology = "MF",
               allGenes = topgo_Np24,
               annot = annFUN.gene2GO,
               geneSel = top5N24_func,
               gene2GO = gene2goMappings)

Np48_go_mf <- new("topGOdata",
               ontology = "MF",
               allGenes = topgo_Np48,
               annot = annFUN.gene2GO,
               geneSel = top5N48_func,
               gene2GO = gene2goMappings)

# creating GO data objects for biological process
Cp24_go_bp <- new("topGOdata", 
               ontology = "BP", 
               allGenes = topgo_Cp24, 
               annot = annFUN.gene2GO, 
               geneSel = top5C24_func,
               gene2GO = gene2goMappings)

Cp48_go_bp <- new("topGOdata",
               ontology = "BP",
               allGenes = topgo_Cp48,
               annot = annFUN.gene2GO,
               geneSel = top5C48_func,
               gene2GO = gene2goMappings)

Np24_go_bp <- new("topGOdata",
               ontology = "BP",
               allGenes = topgo_Np24,
               annot = annFUN.gene2GO,
               geneSel = top5N24_func,
               gene2GO = gene2goMappings)

Np48_go_bp <- new("topGOdata",
               ontology = "BP",
               allGenes = topgo_Np48,
               annot = annFUN.gene2GO,
               geneSel = top5N48_func,
               gene2GO = gene2goMappings)

# creating GO data objects for cellular component
Cp24_go_cc <- new("topGOdata", 
               ontology = "CC", 
               allGenes = topgo_Cp24, 
               annot = annFUN.gene2GO, 
               geneSel = top5C24_func,
               gene2GO = gene2goMappings)

Cp48_go_cc <- new("topGOdata",
               ontology = "CC",
               allGenes = topgo_Cp48,
               annot = annFUN.gene2GO,
               geneSel = top5C48_func,
               gene2GO = gene2goMappings)

Np24_go_cc <- new("topGOdata",
               ontology = "CC",
               allGenes = topgo_Np24,
               annot = annFUN.gene2GO,
               geneSel = top5N24_func,
               gene2GO = gene2goMappings)

Np48_go_cc <- new("topGOdata",
               ontology = "CC",
               allGenes = topgo_Np48,
               annot = annFUN.gene2GO,
               geneSel = top5N48_func,
               gene2GO = gene2goMappings)
```

# Running Fisher's Test for Enrichment and Plotting

I need to read more into this, the vignette talks about the statistics a bit
but it straight up went over my head. Essentially we are finding which gene functions
are over represented given our gene universe. I also need to find a way to make a function for this, it is so annoying to copy and paste and change values all the time.

Comment from biostars for Ks vs weighted fishers
Fisher and ks are just two ways of answering the same question: are the most significant genes enriched for any particular GO term annotations?

Fisher's exact test compares the expected number of significant genes at random to the observed number of significant genes to arrive at a probability.

The KS test compares the distribution of gene p-values expected at random to the observed distribution of the gene p-values to arrive at a probability. KS is theoretically the better choice because it does not require an arbitrary p-value threshold.

Based of of my most recent project, however, the fisher test with p<0.01 and weight01 algorithm seemed to identify the informative GO terms, whereas KS and weight01 tended to identify very basic GO terms like biological process, or cellular process. This could be particular to my dataset, so I would suggest trying both and seeing which gives you more informative GO terms.

```{r, echo=FALSE,message=FALSE}
# Enrichment tests for Molecular Function Fisher
fisher_Cp24_mf <- runTest(Cp24_go_mf, algorithm = "weight01", statistic = "fisher")
fisher_Cp48_mf <- runTest(Cp48_go_mf, algorithm = "weight01", statistic = "fisher")
fisher_Np24_mf <- runTest(Np24_go_mf, algorithm = "weight01", statistic = "fisher")
fisher_Np48_mf <- runTest(Np48_go_mf, algorithm = "weight01", statistic = "fisher")

# Enrichment tests for Biological Process Fisher
fisher_Cp24_bp <- runTest(Cp24_go_bp, algorithm = "weight01", statistic = "fisher")
fisher_Cp48_bp <- runTest(Cp48_go_bp, algorithm = "weight01", statistic = "fisher")
fisher_Np24_bp <- runTest(Np24_go_bp, algorithm = "weight01", statistic = "fisher")
fisher_Np48_bp <- runTest(Np48_go_bp, algorithm = "weight01", statistic = "fisher")

# Enrichment tests for Cellular Component Fisher
fisher_Cp24_cc <- runTest(Cp24_go_cc, algorithm = "weight01", statistic = "fisher")
fisher_Cp48_cc <- runTest(Cp48_go_cc, algorithm = "weight01", statistic = "fisher")
fisher_Np24_cc <- runTest(Np24_go_cc, algorithm = "weight01", statistic = "fisher")
fisher_Np48_cc <- runTest(Np48_go_cc, algorithm = "weight01", statistic = "fisher")

# Enrichment tests for Molecular Function Kolmogrov-Smirnov
ks_Cp24_mf <- runTest(Cp24_go_mf, algorithm = "classic", statistic = "ks")
ks_Cp48_mf <- runTest(Cp48_go_mf, algorithm = "classic", statistic = "ks")
ks_Np24_mf <- runTest(Np24_go_mf, algorithm = "classic", statistic = "ks")
ks_Np48_mf <- runTest(Np48_go_mf, algorithm = "classic", statistic = "ks")

# Enrichment tests for Biological Process Kolmogrov-Smirnov
ks_Cp24_bp <- runTest(Cp24_go_bp, algorithm = "classic", statistic = "ks")
ks_Cp48_bp <- runTest(Cp48_go_bp, algorithm = "classic", statistic = "ks")
ks_Np24_bp <- runTest(Np24_go_bp, algorithm = "classic", statistic = "ks")
ks_Np48_bp <- runTest(Np48_go_bp, algorithm = "classic", statistic = "ks")

# Enrichment tests for Cellular Component Kolmogrov-Smirnov
ks_Cp24_cc <- runTest(Cp24_go_cc, algorithm = "classic", statistic = "ks")
ks_Cp48_cc <- runTest(Cp48_go_cc, algorithm = "classic", statistic = "ks")
ks_Np24_cc <- runTest(Np24_go_cc, algorithm = "classic", statistic = "ks")
ks_Np48_cc <- runTest(Np48_go_cc, algorithm = "classic", statistic = "ks")

# Enrichment tests for Molecular Function Kolmogrov-Smirnov-Elim
ks_elim_Cp24_mf <- runTest(Cp24_go_mf, algorithm = "elim", statistic = "ks")
ks_elim_Cp48_mf <- runTest(Cp48_go_mf, algorithm = "elim", statistic = "ks")
ks_elim_Np24_mf <- runTest(Np24_go_mf, algorithm = "elim", statistic = "ks")
ks_elim_Np48_mf <- runTest(Np48_go_mf, algorithm = "elim", statistic = "ks")

# Enrichment tests for Biological Process Kolmogrov-Smirnov-Elim
ks_elim_Cp24_bp <- runTest(Cp24_go_bp, algorithm = "elim", statistic = "ks")
ks_elim_Cp48_bp <- runTest(Cp48_go_bp, algorithm = "elim", statistic = "ks")
ks_elim_Np24_bp <- runTest(Np24_go_bp, algorithm = "elim", statistic = "ks")
ks_elim_Np48_bp <- runTest(Np48_go_bp, algorithm = "elim", statistic = "ks")

# Enrichment tests for Cellular Component Kolmogrov-Smirnov-Elim
ks_elim_Cp24_cc <- runTest(Cp24_go_cc, algorithm = "elim", statistic = "ks")
ks_elim_Cp48_cc <- runTest(Cp48_go_cc, algorithm = "elim", statistic = "ks")
ks_elim_Np24_cc <- runTest(Np24_go_cc, algorithm = "elim", statistic = "ks")
ks_elim_Np48_cc <- runTest(Np48_go_cc, algorithm = "elim", statistic = "ks")
```

# Generating a Table for Plotting

This function may be broken, need to test it a bit more. I might have to find a different way to do this since the plots look a bit strange.

```{r, echo=FALSE,message=FALSE}
# Function to generate a nicely formatted enrichment table
get_enrich_table <- function(go_obj, fisher_result, ks_result, ks_elim, top_n = 10, total_genes = NULL) {
  tab <- GenTable(go_obj,
                  classicFisher = fisher_result,
                  classicKS = ks_result,
                  elimKS = ks_elim,
                  orderBy = "classicFisher", # orders results by sig p-values
                  ranksOf = "classicFisher",
                  topNodes = top_n) %>%
    mutate(
      pvalue = as.numeric(classicFisher),
      elimKS = as.numeric(elimKS),
      classicKS = as.numeric(classicKS),
      Term = factor(Term, levels = rev(Term)),
      Count = as.numeric(Significant),
      GeneRatio = if (!is.null(total_genes)) Count / total_genes else NA_real_
    ) %>%
    select(Term, Count, GeneRatio, pvalue, everything()) %>%
    arrange(pvalue)

  return(tab)
}

# Molecular Function (MF)
tab_Cp24_mf <- get_enrich_table(Cp24_go_mf, 
                                fisher_Cp24_mf, 
                                ks_Cp24_mf, 
                                ks_elim_Cp24_mf)

tab_Cp48_mf <- get_enrich_table(Cp48_go_mf, 
                                fisher_Cp48_mf, 
                                ks_Cp48_mf, 
                                ks_elim_Cp48_mf)

tab_Np24_mf <- get_enrich_table(Np24_go_mf, 
                                fisher_Np24_mf,
                                ks_Np24_mf,
                                ks_elim_Np24_mf)

tab_Np48_mf <- get_enrich_table(Np48_go_mf, 
                                fisher_Np48_mf,
                                ks_Np48_mf,
                                ks_elim_Np48_mf)

# Biological Process (BP)
tab_Cp24_bp <- get_enrich_table(Cp24_go_bp, 
                                fisher_Cp24_bp,
                                ks_Cp24_bp,
                                ks_elim_Cp24_bp)

tab_Cp48_bp <- get_enrich_table(Cp48_go_bp, 
                                fisher_Cp48_bp,
                                ks_Cp48_bp,
                                ks_elim_Cp48_bp)

tab_Np24_bp <- get_enrich_table(Np24_go_bp, 
                                fisher_Np24_bp,
                                ks_Np24_bp,
                                ks_elim_Np24_bp)

tab_Np48_bp <- get_enrich_table(Np48_go_bp, 
                                fisher_Np48_bp,
                                ks_Np48_bp,
                                ks_elim_Np48_bp)

# Cellular Component (CC)
tab_Cp24_cc <- get_enrich_table(Cp24_go_cc, 
                                fisher_Cp24_cc,
                                ks_Cp24_cc,
                                ks_elim_Cp24_cc)

tab_Cp48_cc <- get_enrich_table(Cp48_go_cc, 
                                fisher_Cp48_cc,
                                ks_Cp48_cc,
                                ks_elim_Cp48_cc)

tab_Np24_cc <- get_enrich_table(Np24_go_cc, 
                                fisher_Np24_cc,
                                ks_Np24_cc,
                                ks_elim_Np24_cc)

tab_Np48_cc <- get_enrich_table(Np48_go_cc, 
                                fisher_Np48_cc,
                                ks_Np48_cc,
                                ks_elim_Np48_cc)
```

# Plotting Bar Plot

Some standard ggplots for enriched GO functions. Tables need to be ordered prior to plotting.

```{r}
# plotting function
plot_enrich_bar <- function(enrich_tab, title = "") {
  enrich_tab <- enrich_tab %>%
    filter(pvalue < 0.05) %>% # filters out non-significant genes
    arrange(pvalue) %>% # arranges bars by p-value
    mutate(
      Term = factor(Term, levels = rev(unique(Term))),  # keep in plotted order
      Count = as.numeric(Count),
      pvalue = as.numeric(pvalue)
    )

  ggplot(enrich_tab, aes(x = Count, y = Term, fill = pvalue)) +
    geom_bar(stat = "identity") +
    scale_fill_gradient(low = "red", high = "blue", name = "p.adjust") +
    labs(
      title = title,
      x = "Gene Count",
      y = NULL
    ) +
    theme_minimal() +
    theme(
      plot.title = element_text(hjust = 0.5, face = "bold"),
      axis.text.y = element_text(size = 10),
      axis.title.x = element_text(size = 12),
      legend.title = element_text(size = 10),
      legend.text = element_text(size = 9)
    )
}

# MF plots
go_MF_graph_C24 <- plot_enrich_bar(tab_Cp24_mf, "MF Cells vs. Control 24H")
go_MF_graph_C48 <- plot_enrich_bar(tab_Cp48_mf, "MF Cells vs. Control 48H")
go_MF_graph_N24 <- plot_enrich_bar(tab_Np24_mf, "MF Supernatant vs. Control 24H")
go_MF_graph_n48 <- plot_enrich_bar(tab_Np48_mf, "MF Supernatant vs. Control 48H")

# BP plots
go_bp_graph_C24 <- plot_enrich_bar(tab_Cp24_bp, "BP Cells vs. Control 24H")
go_bp_graph_C48 <- plot_enrich_bar(tab_Cp48_bp, "BP Cells vs. Control 48H")
go_bp_graph_N24 <- plot_enrich_bar(tab_Np24_bp, "BP Supernatant vs. Control 24H")
go_bp_graph_N48 <- plot_enrich_bar(tab_Np48_bp, "BP Supernatant vs. Control 48H")

# CC plots
go_CC_graph_C24 <- plot_enrich_bar(tab_Cp24_cc, "CC Cells vs. Control 24H")
go_CC_graph_C48 <- plot_enrich_bar(tab_Cp48_cc, "CC Cells vs. Control 48H")
go_CC_graph_N24 <- plot_enrich_bar(tab_Np24_cc, "CC Supernatant vs. Control 24H")
go_CC_graph_N48 <- plot_enrich_bar(tab_Np48_cc, "CC Supernatant vs. Control 48H")

# saving graphs
ggsave("DEG_results/go_MF_C24.png",
       plot = go_MF_graph_C24, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_MF_C48.png",
       plot = go_MF_graph_C48, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_MF_N24.png",
       plot = go_MF_graph_N24, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_MF_N48.png",
       plot = go_MF_graph_n48, dpi = 600, width = 10, height = 8)

ggsave("DEG_results/go_bp_C24.png",
       plot = go_bp_graph_C24, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_bp_C48.png",
       plot = go_bp_graph_C48, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_bp_N24.png",
       plot = go_bp_graph_N24, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_bp_N48.png",
       plot = go_bp_graph_N48, dpi = 600, width = 10, height = 8)

ggsave("DEG_results/go_CC_C24.png",
       plot = go_CC_graph_C24, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_CC_C48.png",
       plot = go_CC_graph_C48, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_CC_N24.png",
       plot = go_CC_graph_N24, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_CC_N48.png",
       plot = go_CC_graph_N48, dpi = 600, width = 10, height = 8)

# plot to markdown
go_MF_graph_C24
go_MF_graph_C48
go_MF_graph_N24
go_MF_graph_n48

go_bp_graph_C24
go_bp_graph_C48
go_bp_graph_N24
go_bp_graph_N48

go_CC_graph_C24
go_CC_graph_C48
go_CC_graph_N24
go_CC_graph_N48
```

# COG Distribution

So kegg pathway enrichment did not work in the slightest, the KEGG terms are not biologically relevant at all for whatever reason, not exactly sure if I can find a different annotation tool and redo the analysis but it is what it is.

Instead I am going to do something a little more basic and plot the distribution of COG terms for the DEGS, this will provide some useful information about the generic processes occurring.

```{r}
# pulling out all upregulated DEGs
upreg_C24 <- resC24_df %>%
  filter(log2FoldChange > 1.5, padj < 0.5)

upreg_C48 <- resC48_df %>%
  filter(log2FoldChange > 1.5, padj < 0.05)

upreg_N24 <- resN24_df %>%
  filter(log2FoldChange > 1.5, padj < 0.05)

upreg_N48 <- resN48_df %>%
  filter(log2FoldChange > 1.5, padj < 0.05)

# same with downreg
downreg_C24 <- resC24_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05)

downreg_C48 <- resC48_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05)

downreg_N24 <- resN24_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05)

downreg_N48 <- resN48_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05)
```

# Adding COG terms

```{r}
upreg_C24 <- merge(upreg_C24, annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)

upreg_C48 <- merge(upreg_C48, annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)

upreg_N24 <- merge(upreg_N24, annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)

upreg_N48 <- merge(upreg_N48, annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)

downreg_C24 <- merge(downreg_C24, annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)

downreg_C48 <- merge(downreg_C48, annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)

downreg_N24 <- merge(downreg_N24,annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)

downreg_N48 <- merge(downreg_N48,annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)
```

# Filter out the non matches

```{r}
upreg_C24 <- upreg_C24 %>%
  filter(COG_category != "-")

upreg_C48 <- upreg_C48 %>%
  filter(COG_category != "-")

upreg_N24 <- upreg_N24 %>%
  filter(COG_category != "-")

upreg_N48 <- upreg_N48 %>%
  filter(COG_category != "-")

downreg_C24 <- downreg_C24 %>%
  filter(COG_category != "-")

downreg_C48 <- downreg_C48 %>%
  filter(COG_category != "-")

downreg_N24 <- downreg_N24 %>%
  filter(COG_category != "-")

downreg_N48 <- downreg_N48 %>%
  filter(COG_category != "-")
```

# Making our Bar plots for upreg COGs

```{r}
# creating our bar plots
cog_bar_c24 <- ggplot(data = upreg_C24, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  scale_fill_viridis_d(option = "C") +
  labs(
    title = "Upregulated COG Categories: Cells vs. Control 24H",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

cog_bar_c48 <- ggplot(data = upreg_C48, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  scale_fill_viridis_d(option = "C") +
  labs(
    title = "Upregulated COG Categories: Cells vs. Control 48H",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

cog_bar_n24 <- ggplot(data = upreg_N24, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  scale_fill_viridis_d(option = "C") +
  labs(
    title = "Upregulated COG Categories: Supernatant vs. Control 24H",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

cog_bar_n48 <- ggplot(data = upreg_N48, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  scale_fill_viridis_d(option = "C") +
  labs(
    title = "Upregulated COG Categories: Supernatant vs. Control 48H",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

# save files
#ggsave("DEG_results/up_cogs_c24.png", plot = cog_bar_c24, dpi = 600, width = 10, height = 8)
#ggsave("DEG_results/up_cogs_c48.png", plot = cog_bar_c48, dpi = 600, width = 10, height = 8)
#ggsave("DEG_results/up_cogs_n24.png", plot = cog_bar_n24, dpi = 600, width = 10, height = 8)
#ggsave("DEG_results/up_cogs_n48.png", plot = cog_bar_n48, dpi = 600, width = 10, height = 8)

# plot to markdown
cog_bar_c24
cog_bar_c48
cog_bar_n24
cog_bar_n48
```
# Saving as multipanel figure

```{r}
cog_first <- plot_grid(cog_bar_c24, cog_bar_c48, cog_bar_n24, cog_bar_n48, 
                             labels = c("A","B","C","D"),
                             ncol = 2,
                             align = "hv")
# save plot
#ggsave("DEG_results/first_cog.pdf", 
       #plot = cog_first, width = 12, height = 8)

# plot to markdown
cog_first
```


# Plotting downreg

```{r}
# creating our bar plots
cog_bar_c24D <- ggplot(data = downreg_C24, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  geom_text(stat='count', aes(label=..count..), vjust=-0.3, size=3.5) +  # add count labels
  scale_fill_viridis_d(option = "D") +
  labs(
    title = "Downregulated COG Categories: Cells vs. Control 24",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

cog_bar_c48D <- ggplot(data = downreg_C48, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  geom_text(stat='count', aes(label=..count..), vjust=-0.3, size=3.5) +  # add count labels
  scale_fill_viridis_d(option = "D") +
  labs(
    title = "Downregulated COG Categories: Cells vs. Control 48",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

cog_bar_n24D <- ggplot(data = downreg_N24, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  geom_text(stat='count', aes(label=..count..), vjust=-0.3, size=3.5) +  # add count labels
  scale_fill_viridis_d(option = "D") +
  labs(
    title = "Downregulated COG Categories: Supernatant vs. Control 24",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

cog_bar_n48D <- ggplot(data = downreg_N48, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  geom_text(stat='count', aes(label=..count..), vjust=-0.3, size=3.5) +  # add count labels
  scale_fill_viridis_d(option = "D") +
  labs(
    title = "Downregulated COG Categories: Supernatant vs. Control 48",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

# save files
ggsave("DEG_results/down_cogs_c24.png",
       plot = cog_bar_c24D, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/down_cogs_c48.png",
       plot = cog_bar_c48D, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/down_cogs_n24.png",
       plot = cog_bar_n24D, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/down_cogs_n48.png",
       plot = cog_bar_n48D, dpi = 600, width = 10, height = 8)

# plot to markdown
cog_bar_c24D
cog_bar_c48D
cog_bar_n24D
cog_bar_n48D
```


# Redo of analysis with different pairwise comparisons

Alrighty so we got our data analysis down pretty pat. Now I want to compare the different time points to each other

```{r}
# extract the comparisons we need for the cells 24 and 48
cells_time_frame <- counts[-c(1,2,3,7,8,9,13,14,15)] # there is probably a better way to do this honestly, im removing columns from the original counts dataframe to leave the ones I am interested in, I bet dplyr has a better way to do this tbh

# extract column data of interest
cells_column_data <- coldata[c(4,5,6,10,11,12),] # the comma denotes rows, without a comma it denotes the columns
```

# DESeq Analysis Formula

```{r}
### DEG ANALYSIS ###

# creates a DeseqDataSet or DDS
cells_dds <- DESeqDataSetFromMatrix(countData = cells_time_frame,
                              colData = cells_column_data,
                              design = ~ Treatment)

# sets the comparisons, in this case the coldata column name is Treatment,
# the reference is named cells48, I am gonna try this and see how it goes
cells_dds$Treatment <- relevel(cells_dds$Treatment, ref = "cells24")

# no clue what this does, manual said to tho
# prints the dataframe i suppose?
cells_dds

# performs DEG
cells_dds <- DESeq(cells_dds)

# creates a results dataframe
cells_results <- results(cells_dds)
```

# Sanity Checks

```{r}
# prints results
cells_results

# prints summary of results
summary(cells_results)

# prints pairwise comparisons,useful double check
resultsNames(cells_dds)
```

# Now lets do the same for the supernatant

```{r}
supernat_time_frame <- counts[-c(1:6,10:12)] # even now i made it more simple but theres gotta be a better way right?

# extract column data of interest
supernat_column_data <- coldata[c(7:9,13:15),] # the comma denotes rows, without a comma it denotes the columns
```

# DESeq Analysis

```{r}
### DEG ANALYSIS ###

# creates a DeseqDataSet or DDS
super_dds <- DESeqDataSetFromMatrix(countData = supernat_time_frame,
                              colData = supernat_column_data,
                              design = ~ Treatment)

# sets the comparisons, in this case the coldata column name is Treatment,
# the reference is named cells48, I am gonna try this and see how it goes
super_dds$Treatment <- relevel(super_dds$Treatment, ref = "NINA24")

# no clue what this does, manual said to tho
# prints the dataframe i suppose?
super_dds

# performs DEG
super_dds <- DESeq(super_dds)

# creates a results dataframe
super_results <- results(super_dds)
```

# Sanity Checks

```{r}
# prints results
super_results

# prints summary of results
summary(super_results)

# prints pairwise comparisons,useful double check
resultsNames(super_dds)
```

# Time for plotting and analysis

A lot of this is boilerplate unfortunately :(

```{r}
# store comparison names
cells_names <- resultsNames(cells_dds)[-c(1)]
super_names <- resultsNames(super_dds)[-c(1)]

#saves results for each comparison
cellsRes <- results(cells_dds, name = cells_names)
superRes <- results(super_dds, name = super_names) 

#filter the data for NA values
cells_df <- as.data.frame(cellsRes) %>% # pipe operator, the same as | foo | in bash
  filter(!is.na(log2FoldChange), !is.na(padj))
super_df <- as.data.frame(superRes) %>%
  filter(!is.na(log2FoldChange), !is.na(padj))

# add columns for up, down, and no-sig
cells_df$diffexpress <- "NO" # initializes column with no as default
cells_df$diffexpress[cells_df$padj < 0.05 & cells_df$log2FoldChange > 1.5] <- "UP"
cells_df$diffexpress[cells_df$padj < 0.05 & cells_df$log2FoldChange < -1.5] <- "DOWN"

super_df$diffexpress <- "NO" # initializes column with no as default
super_df$diffexpress[super_df$padj < 0.05 & super_df$log2FoldChange > 1.5] <- "UP"
super_df$diffexpress[super_df$padj < 0.05 & super_df$log2FoldChange < -1.5] <- "DOWN"

# convert row names to columns
cells_df$gene_id <- rownames(cells_df)
super_df$gene_id <- rownames(super_df)
```

# Volcanos

```{r}
# creates a basic volcano plot as a starting point
cells_volc <- ggplot(data = cells_df, 
                        aes(x = log2FoldChange, 
                            y = -log10(padj), 
                            col = diffexpress)) +
  geom_vline(xintercept = c(-1.5, 1.5), col = "gray", linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = "dashed") +
  geom_point(size = 2) +
  scale_color_manual(values = c("blue","grey","red"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, 75), xlim = c(-10, 10)) +
  labs(color = "Legend", 
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"p-value")) +
  ggtitle("Cells 48 vs. Cells 24") +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

super_volc <- ggplot(data = super_df, 
                        aes(x = log2FoldChange, 
                            y = -log10(padj), 
                            col = diffexpress)) +
  geom_vline(xintercept = c(-1.5, 1.5), col = "gray", linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), col = "gray", linetype = "dashed") +
  geom_point(size = 2) +
  scale_color_manual(values = c("blue","grey","red"),
                     labels = c("Downregulated", "Not significant", "Upregulated")) +
  coord_cartesian(ylim = c(0, 40), xlim = c(-10, 10)) +
  labs(color = "Legend", 
       x = expression("log"[2]*"FC"), 
       y = expression("-log"[10]*"p-value")) +
  ggtitle("Supernatant 48 vs. Supernatant 24") +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

# plot to markdown
cells_volc
super_volc
```
# Combo figure time

```{r}
# removing legend
cells_volc <- cells_volc + theme(legend.position = "none")
super_volc <- super_volc + theme(legend.position = "none")

second_volc <- plot_grid(cells_volc, super_volc,
                         labels = c("A","B"),
                         ncol = 2,
                         align = "hv")
# saving plot
ggsave("DEG_results/second_volc.png", 
       plot = second_volc, dpi = 600, width = 10, height = 8)

second_volc
```


# Adding Diamond annotations

```{r}
# function to add annotations to results dataframe
add_diamond_annotations <- function(res_df, diamond_df) {
  merge(res_df, diamond_df[, c("qseqid", "sseqid", "stitle")],
        by.x = "gene_id", by.y = "qseqid", all.x = TRUE) # uses a left join (all.x)
}

# running our function for each data frame
DIAMOND_cells_df <- add_diamond_annotations(cells_df, diamond_hits)
DIAMOND_super_df <- add_diamond_annotations(super_df, diamond_hits)

# adding a new column with only upregulated genes for labeling
add_delabel_top50 <- function(df) {
  
  # selects our significant genes using both values
  sig_genes <- df %>%
    filter(abs(log2FoldChange) > 1.5, pvalue < 0.05)
  
  # selects the top 25 by p-value
  top50 <- sig_genes %>%
    arrange(pvalue) %>%
    slice_head(n = 10)
  
  # add delabel column, label only top 50 genes
  df$delabel <- ifelse(df$sseqid %in% top50$sseqid, df$sseqid, NA)
  
  return(df)
}

# Apply to your dataframes
DIAMOND_cells_df <- add_delabel_top50(DIAMOND_cells_df)
DIAMOND_super_df <- add_delabel_top50(DIAMOND_super_df)
```

# Go enrichment

```{r}
# making a frame with only the genes with GO terms
go_cells <- merge(cells_df, go_terms[, c("query", "GOs")], 
                by.x = "gene_id", 
                by.y = "query", 
                all.x = TRUE)

go_super <-merge(super_df, go_terms[, c("query", "GOs")], 
               by.x = "gene_id", 
               by.y = "query", 
               all.x = TRUE)

# filter out any genes with no GO term
go_cells <- go_cells[!is.na(go_cells$GOs),]
go_super <- go_super[!is.na(go_super$GOs),]

# filter by log fold change
go_cells <- go_cells %>%
  filter(abs(log2FoldChange) > 1.5)

go_super <- go_super %>%
  filter(abs(log2FoldChange) > 1.5)

# Extract gene IDs and p-values
Cells_GO <- dplyr::select(go_cells, gene_id,padj)
Super_GO <- dplyr::select(go_super, gene_id,padj)

# sort by p-value
sorted_Cells <- Cells_GO[order(go_cells$padj),]
sorted_Super <- Super_GO[order(go_super$padj),]

# convert to topGOs genelist format
topgo_Cells <- sorted_Cells$padj
topgo_Super <- sorted_Super$padj

# name the topgo vector
names(topgo_Cells) <- sorted_Cells$gene_id
names(topgo_Super) <- sorted_Super$gene_id

# filtering value
padj_cut <- 0.05

# functions for filtering significant genes
topCells_func <- function(x) {
  return(x <= padj_cut)
}

topSuper_func <- function(x) {
  return(x <= padj_cut)
}
```

# Go enrichment function

```{r, echo=FALSE, message=FALSE}
# creating GO data objects for molecular function
Cells_go_mf <- new("topGOdata", 
               ontology = "MF", 
               allGenes = topgo_Cells, 
               annot = annFUN.gene2GO, 
               geneSel = topCells_func,
               gene2GO = gene2goMappings)

Super_go_mf <- new("topGOdata",
               ontology = "MF",
               allGenes = topgo_Super,
               annot = annFUN.gene2GO,
               geneSel = topSuper_func,
               gene2GO = gene2goMappings)

# creating GO data objects for biological process
Cells_go_bp <- new("topGOdata", 
               ontology = "BP", 
               allGenes = topgo_Cells, 
               annot = annFUN.gene2GO, 
               geneSel = topCells_func,
               gene2GO = gene2goMappings)

Super_go_bp <- new("topGOdata",
               ontology = "BP",
               allGenes = topgo_Super,
               annot = annFUN.gene2GO,
               geneSel = topSuper_func,
               gene2GO = gene2goMappings)

# creating GO data objects for cellular component
Cells_go_cc <- new("topGOdata", 
               ontology = "CC", 
               allGenes = topgo_Cells, 
               annot = annFUN.gene2GO, 
               geneSel = topCells_func,
               gene2GO = gene2goMappings)

Super_go_cc <- new("topGOdata",
               ontology = "CC",
               allGenes = topgo_Super,
               annot = annFUN.gene2GO,
               geneSel = topSuper_func,
               gene2GO = gene2goMappings)
```
# Fishers test

```{r, echo=FALSE, message=FALSE}
# Enrichment tests for Molecular Function Fisher
fisher_cells_mf <- runTest(Cells_go_mf, algorithm = "weight01", statistic = "fisher")
fisher_super_mf <- runTest(Super_go_mf, algorithm = "weight01", statistic = "fisher")

# Enrichment tests for Biological Process Fisher
fisher_cells_bp <- runTest(Cells_go_bp, algorithm = "weight01", statistic = "fisher")
fisher_super_bp <- runTest(Super_go_bp, algorithm = "weight01", statistic = "fisher")

# Enrichment tests for Cellular Component Fisher
fisher_cells_cc <- runTest(Cells_go_cc, algorithm = "weight01", statistic = "fisher")
fisher_super_cc <- runTest(Super_go_cc, algorithm = "weight01", statistic = "fisher")

# Enrichment tests for Molecular Function Kolmogrov-Smirnov
ks_cells_mf <- runTest(Cells_go_mf, algorithm = "classic", statistic = "ks")
ks_super_mf <- runTest(Super_go_mf, algorithm = "classic", statistic = "ks")

# Enrichment tests for Biological Process Kolmogrov-Smirnov
ks_cells_bp <- runTest(Cells_go_bp, algorithm = "classic", statistic = "ks")
ks_super_bp <- runTest(Super_go_bp, algorithm = "classic", statistic = "ks")

# Enrichment tests for Cellular Component Kolmogrov-Smirnov
ks_cells_cc <- runTest(Cells_go_cc, algorithm = "classic", statistic = "ks")
ks_super_cc <- runTest(Super_go_cc, algorithm = "classic", statistic = "ks")

# Enrichment tests for Molecular Function Kolmogrov-Smirnov-Elim
ks_elim_cells_mf <- runTest(Cells_go_mf, algorithm = "elim", statistic = "ks")
ks_elim_super_mf <- runTest(Super_go_mf, algorithm = "elim", statistic = "ks")

# Enrichment tests for Biological Process Kolmogrov-Smirnov-Elim
ks_elim_cells_bp <- runTest(Cells_go_bp, algorithm = "elim", statistic = "ks")
ks_elim_super_bp <- runTest(Super_go_bp, algorithm = "elim", statistic = "ks")

# Enrichment tests for Cellular Component Kolmogrov-Smirnov-Elim
ks_elim_cells_cc <- runTest(Cells_go_cc, algorithm = "elim", statistic = "ks")
ks_elim_super_cc <- runTest(Super_go_cc, algorithm = "elim", statistic = "ks")
```

# tables

```{r}
# Molecular Function (MF)
tab_cells_mf <- get_enrich_table(Cells_go_mf, 
                                fisher_cells_mf, 
                                ks_cells_mf, 
                                ks_elim_cells_mf)

tab_super_mf <- get_enrich_table(Super_go_mf, 
                                fisher_super_mf, 
                                ks_super_mf, 
                                ks_elim_super_mf)

# Biological Process (BP)
tab_cells_bp <- get_enrich_table(Cells_go_bp, 
                                fisher_cells_bp, 
                                ks_cells_bp, 
                                ks_elim_cells_bp)

tab_super_bp <- get_enrich_table(Super_go_bp, 
                                fisher_super_bp, 
                                ks_super_bp, 
                                ks_elim_super_bp)

# Cellular Component (CC)
tab_cells_cc <- get_enrich_table(Cells_go_cc, 
                                fisher_cells_cc, 
                                ks_cells_cc, 
                                ks_elim_cells_cc)

tab_super_cc <- get_enrich_table(Super_go_cc, 
                                fisher_super_cc, 
                                ks_super_cc, 
                                ks_elim_super_cc)
```

# Plotting

```{r}
# MF plots
go_MF_graph_cells <- plot_enrich_bar(tab_cells_mf, "MF Cells")
go_MF_graph_super <- plot_enrich_bar(tab_super_mf, "MF Supernatant")

# BP plots
go_bp_graph_cells <- plot_enrich_bar(tab_cells_bp, "BP Cells")
go_bp_graph_super <- plot_enrich_bar(tab_super_bp, "BP Supernatant")

# CC plots
go_CC_graph_cells <- plot_enrich_bar(tab_cells_cc, "CC Cells")
go_CC_graph_super <- plot_enrich_bar(tab_super_cc, "CC Supernatant")

# saving graphs
ggsave("DEG_results/go_MF_cells.png",
       plot = go_MF_graph_cells, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_MF_super.png",
       plot = go_MF_graph_super, dpi = 600, width = 10, height = 8)

ggsave("DEG_results/go_bp_cells.png",
       plot = go_bp_graph_cells, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_bp_super.png",
       plot = go_bp_graph_super, dpi = 600, width = 10, height = 8)

ggsave("DEG_results/go_CC_cells.png",
       plot = go_CC_graph_cells, dpi = 600, width = 10, height = 8)
ggsave("DEG_results/go_CC_super.png",
       plot = go_CC_graph_super, dpi = 600, width = 10, height = 8)

# plot to markdown
go_MF_graph_cells
go_MF_graph_super

go_bp_graph_cells
go_bp_graph_super

go_CC_graph_cells
go_CC_graph_super
```

# Cog stuff

```{r}
# pulling out all upregulated DEGs
upreg_cells <- cells_df %>%
  filter(log2FoldChange > 1.5, padj < 0.5)

upreg_super <- super_df %>%
  filter(log2FoldChange > 1.5, padj < 0.05)

# same with downreg
downreg_cells <- cells_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05)

downreg_super <- super_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05)
```

# Adding cog terms

```{r}
upreg_cells <- merge(upreg_cells, annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)

upreg_super <- merge(upreg_super, annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)

downreg_cells <- merge(downreg_cells, annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)

downreg_super <- merge(downreg_super, annotation_data[, c("query", "COG_category")], by.x = "gene_id", by.y = "query", all.x = TRUE)
```

# Filter out non matches

```{r}
upreg_cells <- upreg_cells %>%
  filter(COG_category != "-")

upreg_super <- upreg_super %>%
  filter(COG_category != "-")

downreg_cells <- downreg_cells %>%
  filter(COG_category != "-")

downreg_super <- downreg_super %>%
  filter(COG_category != "-")
```


# bar plotting

```{r}
# creating our bar plots
cog_bar_upcells <- ggplot(data = upreg_cells, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  geom_text(stat='count', aes(label=..count..), vjust=-0.3, size=3.5) +  # add count labels
  scale_fill_viridis_d(option = "C") +
  labs(
    title = "Upregulated COG Categories: Cells 48H vs. 24H",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

cog_bar_upsuper <- ggplot(data = upreg_super, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  geom_text(stat='count', aes(label=..count..), vjust=-0.3, size=3.5) +  # add count labels
  scale_fill_viridis_d(option = "C") +
  labs(
    title = "Upregulated COG Categories: Supernatant 48H vs. 24H",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

# creating our bar plots
cog_bar_downcells <- ggplot(data = downreg_cells, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  geom_text(stat='count', aes(label=..count..), vjust=-0.3, size=3.5) +  # add count labels
  scale_fill_viridis_d(option = "D") +
  labs(
    title = "Downregulated COG Categories: Cells 48H vs. 24H",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

cog_bar_downsuper <- ggplot(data = downreg_super, aes(x = COG_category, fill = COG_category)) + 
  geom_bar(show.legend = FALSE) +  # remove legend for cleaner look
  geom_text(stat='count', aes(label=..count..), vjust=-0.3, size=3.5) +  # add count labels
  scale_fill_viridis_d(option = "D") +
  labs(
    title = "Upregulated COG Categories: Supernatant 48H vs. 24H",
    x = "COG Category",
    y = "Category Count"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, face = "bold"),
    axis.text.y = element_text(face = "bold"),
    axis.title = element_text(face = "bold"),
    plot.title = element_text(face = "bold", hjust = 0.5)
  )

# save pngs

# plot to markdown
cog_bar_upcells
cog_bar_upsuper
cog_bar_downcells
cog_bar_downsuper
```


# Export out top genes for everythinggg

```{r}
# top 20 genes
C24_top20 <- DIAMOND_resC24_df %>%
  filter(log2FoldChange > 1.5, padj < 0.05) %>%
  arrange(desc(log2FoldChange)) %>%
  slice_head(n = 20)

C48_top20 <- DIAMOND_resC48_df %>%
  filter(log2FoldChange > 1.5, padj < 0.05) %>%
  arrange(desc(log2FoldChange)) %>%
  slice_head(n = 20)

N24_top20 <- DIAMOND_resN24_df %>%
  filter(log2FoldChange > 1.5, padj < 0.05) %>%
  arrange(desc(log2FoldChange)) %>%
  slice_head(n = 20)

N48_top20 <- DIAMOND_resN48_df %>%
  filter(log2FoldChange > 1.5, padj < 0.05) %>%
  arrange(desc(log2FoldChange)) %>%
  slice_head(n = 20)

cells_top20 <- DIAMOND_cells_df %>%
  filter(log2FoldChange > 1.5, padj < 0.05) %>%
  arrange(desc(log2FoldChange)) %>%
  slice_head(n = 20)

super_top20 <- DIAMOND_super_df %>%
  filter(log2FoldChange > 1.5, padj < 0.05) %>%
  arrange(desc(log2FoldChange)) %>%
  slice_head(n = 20)

# bottom 20 genes
C24_bott20 <- DIAMOND_resC24_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05) %>%
  arrange(log2FoldChange) %>%
  slice_head(n = 20)

C48_bott20 <- DIAMOND_resC48_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05) %>%
  arrange(log2FoldChange) %>%
  slice_head(n = 20)

N24_bott20 <- DIAMOND_resN24_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05) %>%
  arrange(log2FoldChange) %>%
  slice_head(n = 20)

N48_bott20 <- DIAMOND_resN48_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05) %>%
  arrange(log2FoldChange) %>%
  slice_head(n = 20)

cells_bott20 <- DIAMOND_cells_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05) %>%
  arrange(log2FoldChange) %>%
  slice_head(n = 20)

super_bott20 <- DIAMOND_super_df %>%
  filter(log2FoldChange < -1.5, padj < 0.05) %>%
  arrange(log2FoldChange) %>%
  slice_head(n = 20)
```

# Gonna try to make some nice tables

This one has colors
gt_C24 <- C24_top20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Upregulated Genes"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  ) %>%
  data_color(
    columns = log2FoldChange,
    colors = scales::col_numeric(c("orange", "red"), domain = NULL)
  )

```{r}
# top 20 tables
gt_C24 <- C24_top20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Upregulated Genes: Cells 24H vs. Control"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

gt_C48 <- C48_top20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Upregulated Genes: Cells 48H vs. Control"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

gt_N24 <- N24_top20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Upregulated Genes: Supernatant 24H vs. Control"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

gt_N48 <- N48_top20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top Upregulated Genes: Supernatant 48H vs. Control"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

gt_cells <- cells_top20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Upregulated Genes: Cells 48H vs Cells 24H"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

gt_super <- super_top20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top Upregulated Genes: Supernatant 48H vs. Supernatant 24H"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

# save table
#gtsave(gt_C24, "C24_table.png")
gt_C24
gt_C48
gt_N24
gt_N48
gt_cells
gt_super
```
# Downreg tables

```{r}
# downregulated genes
gt_C24D <- C24_bott20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Downregulated Genes: Cells 24H vs. Control"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

gt_C48D <- C48_bott20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Downregulated Genes: Cells 48H vs. Control"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

gt_N24D <- N24_bott20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Downregulated Genes: Supernatant 24H vs. Control"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

gt_N48D <- N48_bott20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Downregulated Genes: Supernatant 48H vs. Control"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

gt_cellsD <- cells_bott20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Downregulated Genes: Cells 48H vs Cells 24H"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

gt_superD <- super_bott20 %>%
  select(gene_id, log2FoldChange, padj, stitle) %>%
  gt() %>%
  tab_header(
    title = "Top 20 Downregulated Genes: Supernatant 48H vs. Supernatant 24H"
  ) %>%
  cols_label(
    gene_id = "Gene",
    log2FoldChange = "Log2 Fold Change",
    padj = "Adjusted P-Value",
    stitle = "DIAMOND Hit"
  )

# plot to markdown
gt_C24D
gt_C48D
gt_N24D
gt_N48D
gt_cellsD
gt_superD
```

# Save the tables

```{r, echo=FALSE,message=FALSE}
gtsave(gt_C24, "DEG_results/upregC24.png")
gtsave(gt_C48, "DEG_results/upregC48.png")
gtsave(gt_N24, "DEG_results/upregSuper24.png")
gtsave(gt_N48, "DEG_results/upregSuper48.png")

gtsave(gt_C24D, "DEG_results/downregC24.png")
gtsave(gt_C48D, "DEG_results/downregC48.png")
gtsave(gt_N24D, "DEG_results/downregN24.png")
gtsave(gt_N48D, "DEG_results/downregN48.png")

gtsave(gt_cells, "DEG_results/upregCells.png")
gtsave(gt_super, "DEG_results/upregsuper.png")

gtsave(gt_cellsD, "DEG_results/downregcells.png")
gtsave(gt_superD, "DEG_results/downregsuper.png")
```

